src/Processors/QueryPlan/Optimizations/convertOuterJoinToInnerJoin.cpp

```diff
@@ -99,14 +99,55 @@ size_t tryConvertAnyOuterJoinToInnerJoin(
     if (result_for_not_matched_rows != FilterResult::FALSE)
         return 0;
 
-    auto key_dags = join_step->preCalculateKeys(join_node->children.front()->step->getOutputHeader(), join_node->children.back()->step->getOutputHeader());
-    if (!key_dags)
-        return 0;
+    NameSet join_keys_interesting_side;
+
+    const auto add_key_recursively = [&](const ActionsDAG::Node * node) -> void
+    {
+        if (node->type == ActionsDAG::ActionType::FUNCTION)
+        {
+            std::stack<const ActionsDAG::Node *> nodes_to_process;
+            nodes_to_process.push(node);
+            while (!nodes_to_process.empty())
+            {
+                const auto * current_node = nodes_to_process.top();
+                nodes_to_process.pop();
+
+                join_keys_interesting_side.insert(current_node->result_name);
```

Comment:
I think we should add only `INPUT` nodes, shouldn't we? Example from test: `u1.name = reverse(u2.name)`.
It would contain: `reverse(u2.name)`, `u2.name`.
Would it be enough to have only `u2.name`?

Reply:
I think you are right, it is enough to add `INPUT` nodes. My reasoning is if the node is not an input, then it is not in the grouping key for the group by. There might be a situation (`JoinStepLogical` -> `ExpressionStep` -> `Aggregating`) where the `ExpressionStep` might discard some columns, but I doubt we would discard a column that is used in the join. If this is the case, we can improve this later.